File: C++/CompositionDesignPatternDemo.cpp
#include <iostream>
#include <memory>
#include <string>
#include <unordered_map>

struct IFileSystemNode {
    virtual int getSize() = 0;
    virtual void rename(const std::string& name) = 0;
    virtual void printDetails() = 0;
    virtual ~IFileSystemNode() = default;
};

class File : public IFileSystemNode {
public:
    std::string name;
    int size;
    std::string content;

    File(std::string name, int size) : name(std::move(name)), size(size) {}

    int getSize() override {
        return size;
    }

    void rename(const std::string& newName) override {
        name = newName;
    }

    void printDetails() override {
        std::cout << "File: " << name << std::endl;
    }
};

class Folder : public IFileSystemNode {
public:
    std::string name;
    std::unordered_map<std::string, std::shared_ptr<IFileSystemNode>> childs;

    explicit Folder(std::string name) : name(std::move(name)) {}

    void mkdir(const std::string& dirName) {
        childs[dirName] = std::make_shared<Folder>(dirName);
    }

    void touch(const std::string& fileName, int size) {
        childs[fileName] = std::make_shared<File>(fileName, size);
    }

    void deleteFileSystemNode(const std::string& nodeName) {
        childs.erase(nodeName);
    }

    std::shared_ptr<IFileSystemNode> getChild(const std::string& childName) {
        auto it = childs.find(childName);
        return it != childs.end() ? it->second : nullptr;
    }

    int getSize() override {
        int total = 0;
        for (auto& [childName, node] : childs) {
            total += node->getSize();
        }
        return total;
    }

    void rename(const std::string& newName) override {
        std::cout << "renamed" << std::endl;
    }

    void printDetails() override {
        std::cout << "Current Folder: " << name << std::endl;
        for (auto& [childName, node] : childs) {
            node->printDetails();
        }
    }
};

int main() {
    auto root = std::make_shared<Folder>("root");
    root->mkdir("movies");

    return 0;
}

File: Golang/CompositionDesignPatternDemo.go
package main

import "fmt"

type IFileSystemNode interface {
	GetSize() int
	Rename(name string)
	PrintDetails()
}

type File struct {
	Name    string
	Size    int
	Content string
}

func NewFile(name string, size int) *File {
	return &File{Name: name, Size: size}
}

func (f *File) GetSize() int {
	return f.Size
}

func (f *File) Rename(name string) {
	f.Name = name
}

func (f *File) PrintDetails() {
	fmt.Println("File:", f.Name)
}

type Folder struct {
	Name   string
	Childs map[string]IFileSystemNode
}

func NewFolder(name string) *Folder {
	return &Folder{Name: name, Childs: make(map[string]IFileSystemNode)}
}

func (fo *Folder) Mkdir(name string) {
	fo.Childs[name] = NewFolder(name)
}

func (fo *Folder) Touch(name string, size int) {
	fo.Childs[name] = NewFile(name, size)
}

func (fo *Folder) DeleteFileSystemNode(name string) {
	delete(fo.Childs, name)
}

func (fo *Folder) GetChild(name string) IFileSystemNode {
	return fo.Childs[name]
}

func (fo *Folder) GetSize() int {
	size := 0
	for _, node := range fo.Childs {
		size += node.GetSize()
	}
	return size
}

func (fo *Folder) Rename(name string) {
	fmt.Println("renamed")
}

func (fo *Folder) PrintDetails() {
	fmt.Println("Current Folder:", fo.Name)
	for _, node := range fo.Childs {
		node.PrintDetails()
	}
}

func main() {
	root := NewFolder("root")
	root.Mkdir("movies")
}
